### PR TITLE
Nullable annotate System.Linq.Queryable

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Contributes to https://github.com/dotnet/corefx/issues/40623

### CODE REVIEW

File: src/libraries/System.Linq.Queryable/ref/System.Linq.Queryable.cs
Comment (buyaa-n) [inline src/libraries/System.Linq.Queryable/ref/System.Linq.Queryable.cs]: Why do we need this update?

File: src/libraries/System.Linq.Queryable/ref/System.Linq.Queryable.cs
Comment (buyaa-n) [inline src/libraries/System.Linq.Queryable/ref/System.Linq.Queryable.cs]: Are all API's implementation was not exist within the project or? I think ref assemblies missing some annotation

File: src/libraries/System.Linq.Queryable/ref/System.Linq.Queryable.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Linq.Queryable/ref/System.Linq.Queryable.cs]: This as all generated code. I'm assuming what happened here is somebody changed the type argument name but didn't regenerate the ref sources. I've seen instances of this elsewhere too.

File: src/libraries/System.Linq.Queryable/ref/System.Linq.Queryable.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Linq.Queryable/ref/System.Linq.Queryable.cs]: I didn't see any types requiring nullable annotations leaking out in the public signatures.

File: src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableQuery.cs
Comment (buyaa-n) [inline src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableQuery.cs]: __expression.ToString() wouldn't return null (I am still working on Linq.Expression)

File: src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableQuery.cs
Comment (buyaa-n) [inline src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableQuery.cs]: We normally update such `as` expression to `is` 
`if ( _expression is ConstantExpression c && c.Value == this)`

File: src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableQuery.cs
Comment (buyaa-n) [reply src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableQuery.cs]: please disregard above comment, seems you make it nullable because of _enumerable.ToString() 

_enumerable.ToString()  is calling object.ToString() which is calling Type.ToString() which will not be nullable. (We made object.toString() nullable because it is abstract and there can be overrides returning null )

File: src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableRewriter.cs
Comment (buyaa-n) [inline src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableRewriter.cs]: Please use `is` instead of `as`

File: src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableQuery.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableQuery.cs]: Sure, I'm wondering though whether it's appropriate for the scope of annotation. Wouldn't we generally want to avoid changes that could potentially result in slightly different IL?

Comment (buyaa-n) [reply src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableQuery.cs]: This is one kind of comments I have had before in the scope of annotations PR. Anyways by my understanding  our direction is not to increase IL size, and I assume it wouldn't increase IL size.

Personally I don't like to do any other refactoring with annotations though.

File: src/libraries/System.Linq.Queryable/src/System/Linq/TypeHelper.cs
Comment (buyaa-n) [inline src/libraries/System.Linq.Queryable/src/System/Linq/TypeHelper.cs]: Doesn't really understood why we need `[NotNull]`attribute

File: src/libraries/System.Linq.Queryable/src/System/Linq/Queryable.cs
Comment (buyaa-n) [inline src/libraries/System.Linq.Queryable/src/System/Linq/Queryable.cs]: Seems lots of implementations here depend on Linq.Expression, as disscussed offline gonna wait for that one annotated first

File: src/libraries/System.Linq.Queryable/src/System/Linq/TypeHelper.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Linq.Queryable/src/System/Linq/TypeHelper.cs]: The contract for the `type` argument is not null, however the method body will assign null values to it. The attribute compromises the two without having to declare a new local variable.

**Comment by eiriktsarpalis** [comment]
I wasn't aware that System.Linq.Expressions hadn't been annotated yet, holding off this PR until that work is complete.

File: src/libraries/System.Linq.Queryable/ref/System.Linq.Queryable.cs
Comment (stephentoub) [inline src/libraries/System.Linq.Queryable/ref/System.Linq.Queryable.cs]: It looks like you didn't annotate Queryable... I'd expect most of the annotations/attributes we have in System.Linq's Enumerable to also show up here.  For example, `[return: MaybeNull]` on the XxOrDefault methods that could end up returning null for reference types.  Please go through the LINQ surface area that's been annotated and confirm all annotations and attributes there are appropriately mirrored here.

File: src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableQuery.cs
Comment (stephentoub) [inline src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableQuery.cs]: Shouldn't this be:
```C#
(IEnumerable<ParameterExpression>?)null
```
instead of:
```C#
(IEnumerable<ParameterExpression>)null!
```
?

File: src/libraries/System.Linq.Queryable/src/System/Linq/TypeHelper.cs
Comment (stephentoub) [reply src/libraries/System.Linq.Queryable/src/System/Linq/TypeHelper.cs]: I understand the logic, but I suggest removing the [NotNull].  The implementation works fine if null is passed in, this is internal so there's no documented contract it needs to expose, and the combination of [NotNull] and ? just looks strange and leads to questions.

File: src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableExecutor.cs
Comment (stephentoub) [inline src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableExecutor.cs]: System.Linq.Expressions hasn't been annotated yet.  There's an open PR for it.  I fully expect that once that goes in, this will need to be revisited.  Please do not merge this PR until after that's in (in general we should avoid working on libraries that depend on others that haven't been annotated... this looks like a mistake I contributed to, in that my tool which analyzed the dependencies in https://github.com/dotnet/corefx/issues/40623 apparently missed this one... sorry about that).

File: src/libraries/System.Linq.Queryable/ref/System.Linq.Queryable.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Linq.Queryable/ref/System.Linq.Queryable.cs]: I'm aware. Holding off this PR until Expressions are annotated.

Comment (stephentoub) [reply src/libraries/System.Linq.Queryable/ref/System.Linq.Queryable.cs]: Cool. Thanks.

Comment (stephentoub) [reply src/libraries/System.Linq.Queryable/ref/System.Linq.Queryable.cs]: (Sorry, just saw your PR comment to this affect earlier.)

**Comment by eiriktsarpalis** [comment]
Rebased over latest changes, addressed feedback. @buyaa-n I also had to make one minor change in System.Linq.Expressions nullable annotations.

File: src/libraries/System.Linq.Expressions/ref/System.Linq.Expressions.cs
Comment (stephentoub) [inline src/libraries/System.Linq.Expressions/ref/System.Linq.Expressions.cs]: Why did this change?

File: src/libraries/System.Linq.Expressions/ref/System.Linq.Expressions.cs
Comment (stephentoub) [inline src/libraries/System.Linq.Expressions/ref/System.Linq.Expressions.cs]: This was just broken before?

File: src/libraries/System.Linq.Expressions/ref/System.Linq.Expressions.cs
Comment (stephentoub) [inline src/libraries/System.Linq.Expressions/ref/System.Linq.Expressions.cs]: Why are we removing this?

File: src/libraries/System.Linq.Expressions/ref/System.Linq.Expressions.cs
Comment (stephentoub) [inline src/libraries/System.Linq.Expressions/ref/System.Linq.Expressions.cs]: It looks like desktop has this as an explicit implementation.  But looking back at the .NET Core history, it came into .NET Core as an implicit implementation.  Strange.

File: src/libraries/System.Linq.Expressions/ref/System.Linq.Expressions.cs
Comment (stephentoub) [inline src/libraries/System.Linq.Expressions/ref/System.Linq.Expressions.cs]: There's a ton of change being made to System.Linq.Expressions.cs here, effectively adding / changing public surface area.  Seems like that really should be in its own PR, with the appropriate discussion around it.  Unless I'm missing something?

**Comment by stephentoub** [comment]
> Rebased over latest changes, addressed feedback

Still not ready for review, though, right? i.e. I don't see Queryable updated to match the LINQ APIs.

File: src/libraries/System.Linq.Expressions/ref/System.Linq.Expressions.cs
Comment (buyaa-n) [reply src/libraries/System.Linq.Expressions/ref/System.Linq.Expressions.cs]: The GenerateReferenceSource tool was adding these classes from source, also it was removing/changing some accessors/modifiers. For me as I don't know if they should be added or not I reverted all change except nullable related ones. Seems most of changes here related to that

Comment (eiriktsarpalis) [reply src/libraries/System.Linq.Expressions/ref/System.Linq.Expressions.cs]: As @buyaa-n points out, the changes stem from regenerating sources after making [a single change](https://github.com/dotnet/runtime/pull/979/files/dc6b2e31b84d1ee360b5c72ce6897889e7b85505#diff-93045c934a6feac07e6f303bc1596c12R78) in the System.Linq.Expressions implementation. This shouldn't happen, the expectation is that regenerating reference sources without implementations should not produce a diff. This can be controlled by blacklisting members, c.f. https://github.com/dotnet/runtime/blob/master/src/libraries/System.Reflection.Metadata/ref/ReferenceSourceExcludeApi.txt. I think we should probably revisit in a separate PR.

Comment (stephentoub) [reply src/libraries/System.Linq.Expressions/ref/System.Linq.Expressions.cs]: Thanks. Yes, please revert all changes not relevant to nullable annotations and open an issue about the differences.

Comment (eiriktsarpalis) [reply src/libraries/System.Linq.Expressions/ref/System.Linq.Expressions.cs]: Done

File: src/libraries/System.Linq.Queryable/ref/System.Linq.Queryable.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Linq.Queryable/ref/System.Linq.Queryable.cs]: @stephentoub Would changing type variable names be considered a breaking change here? If that's the case then I should maybe rename `TElement` to `S` in the implementation.

Comment (stephentoub) [reply src/libraries/System.Linq.Queryable/ref/System.Linq.Queryable.cs]: > Would changing type variable names be considered a breaking change here?

Technically, yes, in that it could impact some reflection-based scenarios.

However, the chance of this breaking anyone is very low.  This is an explicit interface implementation, so the only way someone would be impacted by this is if they were using reflection.  Further, for this `S` to make a difference, they would need to be reflecting over the reference assembly rather than over the implementation assembly, which is unlikely to be done by anything where such a difference would be meaningful.  And, if the difference is actually meaningful, it's worth calling out that neither S nor TElement actually matches what's in the interface, which is TResult:
https://github.com/dotnet/runtime/blob/14db363c4acf2f5ce3ed81119af3ba81454c80ae/src/libraries/System.Linq.Expressions/ref/System.Linq.Expressions.cs#L252
:smile:

cc: @cston

File: src/libraries/System.Linq.Expressions/ref/System.Linq.Expressions.cs
Comment (stephentoub) [reply src/libraries/System.Linq.Expressions/ref/System.Linq.Expressions.cs]: Thanks.

**Comment by eiriktsarpalis** [comment]
> I don't see Queryable updated to match the LINQ APIs.

I presume you mean making the queryable extension methods annotation compatible with the [IEnumerable ones](https://github.com/dotnet/runtime/blob/85033887679144f95bd1583fc87249cc2519e6dc/src/libraries/System.Linq/ref/System.Linq.cs#L55-L56)?

**Comment by stephentoub** [comment]
Yes.

**Comment by eiriktsarpalis** [comment]
Done.

File: src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableQuery.cs
Comment (buyaa-n) [inline src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableQuery.cs]: I am still thinking this should return non null? Wondering what am I missing ...

File: src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableRewriter.cs
Comment (buyaa-n) [inline src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableRewriter.cs]: Visit method annotated with [return NotNullIfNotNull("node")], as m.Object is not null here returning expression is not null. Then we will not need `!` below for the `obj`

File: src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableRewriter.cs
Comment (buyaa-n) [inline src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableRewriter.cs]: NIT: seems return type doesn't have to be nullable, then we wouldn't need `[return: NotNullIfNotNull("node")]`

File: src/libraries/System.Linq.Queryable/src/System/Linq/Queryable.cs
Comment (buyaa-n) [inline src/libraries/System.Linq.Queryable/src/System/Linq/Queryable.cs]: NIT: could use is instead

**Review by buyaa-n — APPROVED** [review]
A few question, overall LGTM, thanks!

File: src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableRewriter.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableRewriter.cs]: It's theoretically possible for `m.Object` to be null, for example if `m` is of runtime type `MethodCallExpression`, for which the property [always evaluates to `null`](https://github.com/dotnet/runtime/blob/2c54bd95ab18962fa6a2c51ef9871d8b6f57d91b/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/MethodCallExpression.cs#L25). Even if it is the case that we only consume proper subtypes which are guaranteed to return non-null values, I think it would make not to let that invariant influence annotations in this context, particularly since it doesn't really impact public APIs.

File: src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableQuery.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableQuery.cs]: I wouldn't think so, `_enumerable` can be an arbitrary IEnumerable implementation which can be passed from a public constructor.

**Review by cston — APPROVED** [review]

**Pull request merged**
